server/leaderboard/views.py

```diff
@@ -88,3 +97,160 @@ def get_queryset(self):
+        return channel_configs.get(channel_id)
+
+    def _handle_stats(self, user_id, channel_config, action='increment'):
+        model = channel_config['model']
+        position_type = channel_config['name']
+
+        try:
+            stats = model.objects.select_for_update().get(user_id=user_id)
+
+            if action == 'increment':
+                stats.applied = F('applied') + 1
+            elif action == 'decrement' and stats.applied > 0:
+                stats.applied = F('applied') - 1
+            else:
+                return False
+
+            stats.save()
+            stats.refresh_from_db()
+
+            log_msg = (f"User {user_id} has {stats.applied} {position_type} "
+                      f"applications after {action}")
+            logger.info(log_msg)
+            return True
+
+        except model.DoesNotExist:
+            if action == 'increment':
+                model.objects.create(user_id=user_id, applied=1)
+                logger.info(f"Created new {position_type} stats for user {user_id}")
+                return True
+            return False
+
+    @transaction.atomic
+    def post(self, request, *args, **kwargs):
+        discord_id = request.data.get("discord_id")
+        channel_id = request.data.get("channel_id")
+        emoji = request.data.get("emoji")
+
+        try:
+            user_id = self._get_user_id(discord_id)
+            channel_config = self._get_channel_config(channel_id)
+
+            if not channel_config:
+                logger.warning(
+                    f"Invalid channel_id: {channel_id}, only supports "
+                    f"{INTERNSHIP_CHANNEL_ID} and {NEW_GRAD_CHANNEL_ID}"
+                )
+                return Response(status=status.HTTP_304_NOT_MODIFIED)
+
+            self._handle_stats(user_id, channel_config, 'increment')
+            return Response(status=status.HTTP_201_CREATED)
+
+        except Http404:
+            logger.error(f"User not found for discord_id: {discord_id}")
+            return Response(status=status.HTTP_404_NOT_FOUND)
+
+    @transaction.atomic
+    def delete(self, request, *args, **kwargs):
+        discord_id = request.data.get("discord_id")
+        channel_id = request.data.get("channel_id")
+
+        try:
+            user_id = self._get_user_id(discord_id)
+            channel_config = self._get_channel_config(channel_id)
+
+            if not channel_config:
+                logger.warning(
+                    f"Invalid channel_id: {channel_id}, only supports "
+                    f"{INTERNSHIP_CHANNEL_ID} and {NEW_GRAD_CHANNEL_ID}"
+                )
+                return Response(status=status.HTTP_304_NOT_MODIFIED)
+
+            self._handle_stats(user_id, channel_config, 'decrement')
+            return Response(status=status.HTTP_204_NO_CONTENT)
+
+        except Http404:
+            logger.error(f"User not found for discord_id: {discord_id}")
+            return Response(status=status.HTTP_404_NOT_FOUND)
```

Comment:
Is 404 the right error code to use here?

Reply:
like semantically? or are you worried about errors being returned when people without accounts react

Reply:
I was worried about it just semantically, but now I think it's probably perfectly fine to return a 404 for a missing resource